File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/movie/model/MovieReleaseType.java
package dev.reuss.tmdb.domain.movie.model;

/**
 * Movie release type.
 */
public enum MovieReleaseType {

    PREMIERE(1),
    THEATRICAL_LIMITED(2),
    THEATRICAL(3),
    DIGITAL(4),
    PHYSICAL(5),
    TV(6);

    private final int value;

    MovieReleaseType(int value) {
        this.value = value;
    }

    public int value() {
        return value;
    }

    public static MovieReleaseType fromValue(int value) {
        for (MovieReleaseType type : values()) {
            if (type.value == value) {
                return type;
            }
        }

        throw new IllegalArgumentException("Unknown movie release type: " + value);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/movie/model/MovieReleaseType.kt
package dev.reuss.tmdb.domain.movie.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonValue

/**
 * Movie release type.
 *
 * @property value TMDB release type value
 */
enum class MovieReleaseType(
    @get:JsonValue
    @all:JsonProperty("value")
    val value: Int,
) {
    PREMIERE(1),
    THEATRICAL_LIMITED(2),
    THEATRICAL(3),
    DIGITAL(4),
    PHYSICAL(5),
    TV(6),
    ;

    companion object {
        @JvmStatic
        @JsonCreator(mode = JsonCreator.Mode.DELEGATING)
        fun fromValue(value: Int): MovieReleaseType =
            entries.firstOrNull { it.value == value }
                ?: throw IllegalArgumentException(
                    "Unknown movie release type: $value",
                )
    }
}
